File: minimumFallingSum.cpp
// Time Complexity : O(N)
// Space Complexity : O(N*N)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach
class Solution {
public:
    int minFallingPathSum(vector<vector<int>>& a) {
      int row=a.size();
      int col=a[0].size();
        
      int dp[row][col];
        
      for (int i=0;i<col;i++) {
            dp[0][i]=a[0][i]; // set the first row equal to the first row of the original matrix
        }
        
         for (int i=1;i<row;i++) { // index starts from row2 int the original matrix
            for (int j=0;j<col;j++) {
                if (j==0) { //first column condition
                    dp[i][j]=a[i][j]+ min(dp[i-1][j],dp[i-1][j+1]);
                } else if (j==col-1) { //last column condition
                    dp[i][j]=a[i][j]+ min(dp[i-1][j],dp[i-1][j-1]);
                } else { //rest of the columns
                    dp[i][j]=a[i][j]+ min( min(dp[i-1][j],dp[i-1][j+1]),dp[i-1][j-1]);
                }
            }
        }
        int mini=INT_MAX;
        for(int i=0;i<col;i++){ //loop through the last row to get the min value
            mini=min(mini,dp[row-1][i]);
        }
        // for(int i=0;i<row;i++) {
        //     for(int j=0;j<col;j++){
        //         cout<<dp[i][j]<<" ";
        //     }
        //     cout<<endl;
        // }
        return mini;
    }
};

File: minimum_falling_sum.py
# Time Complexity : O(n*m) where n is the number of rows and m is the number of columns
# Space Complexity : O(1) since we are modifying the input matrix
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : No


# Your code here along with comments explaining your approach in three sentences only
"""
This code solves the problem by starting from 
falling sum for each cell and then iterating to the top row.
"""

class Solution:
    def minFallingPathSum(self, matrix: List[List[int]]) -> int:
        if matrix == None or len(matrix) == 0:
            return  0

        if len(matrix) == 1:
            return min(matrix[0])
        
        for i in range(len(matrix)-2, -1, -1):
            for j in range(len(matrix[0])-1, -1, -1):
                left = sys.maxsize if j == 0 else matrix[i][j] + matrix[i+1][j-1]
                mid = matrix[i][j] + matrix[i+1][j]
                right = sys.maxsize if j == len(matrix[0]) - 1 else matrix[i][j] + matrix[i+1][j+1]

                matrix[i][j] = min(left, mid, right)

        minsum = sys.maxsize
        for i in range(len(matrix)):
            minsum = min(minsum, matrix[0][i])

        return minsum